crates/fuel-core/src/state.rs

```diff
@@ -13,49 +13,98 @@ use std::{
     sync::Arc,
 };
 
-pub type DataSource = Arc<dyn TransactableStorage>;
+pub type DataSource = Arc<dyn TransactableStorage<Column = Column>>;
 pub type Value = Arc<Vec<u8>>;
 pub type KVItem = DatabaseResult<(Vec<u8>, Value)>;
 
+/// A column of the storage.
+pub trait StorageColumn: Clone {
+    /// Returns the name of the column.
+    fn name(&self) -> &'static str;
+
+    /// Returns the id of the column.
+    fn id(&self) -> u32;
+}
+
 pub trait KeyValueStore {
-    fn put(
+    /// The type of the column.
+    type Column: StorageColumn;
+
+    /// Inserts the `Value` into the storage.
+    fn put(&self, key: &[u8], column: Self::Column, value: Value) -> DatabaseResult<()> {
+        self.write(key, column, value.as_ref()).map(|_| ())
+    }
+
+    /// Put the `Value` into the storage and return the old value.
+    fn replace(
         &self,
         key: &[u8],
-        column: Column,
+        column: Self::Column,
         value: Value,
-    ) -> DatabaseResult<Option<Value>>;
-
-    fn write(&self, key: &[u8], column: Column, buf: &[u8]) -> DatabaseResult<usize>;
+    ) -> DatabaseResult<Option<Value>> {
+        // FIXME: This is a race condition. We should use a transaction.
```

Comment:
I moved it here from RocksDB implementation to remember that it really may hurt because it is not an atomic operation, and if someone modifies the database at the same time, the result can be wrong.

Reply:
We should have an issue for that, so we'll remember to fix it.